runtime/runtime-core/native/src/aws/smithy/kotlin/runtime/content/BigDecimalNative.kt

```diff
@@ -4,46 +4,53 @@
  */
 package aws.smithy.kotlin.runtime.content
 
-public actual class BigDecimal actual constructor(value: String) :
+import com.ionspin.kotlin.bignum.decimal.BigDecimal as IonSpinBigDecimal
+
+public actual class BigDecimal private constructor(private val delegate: IonSpinBigDecimal) :
     Number(),
     Comparable<BigDecimal> {
-    public actual fun toPlainString(): String {
-        TODO("Not yet implemented")
-    }
 
-    actual override fun toByte(): Byte {
-        TODO("Not yet implemented")
+    private companion object {
+        /**
+         * Returns a new or existing [BigDecimal] wrapper for the given delegate [value]
+         * @param value The delegate value to wrap
+         * @param left A candidate wrapper which may already contain [value]
+         * @param right A candidate wrapper which may already contain [value]
+         */
+        fun coalesceOrCreate(value: IonSpinBigDecimal, left: BigDecimal, right: BigDecimal): BigDecimal = when (value) {
+            left.delegate -> left
+            right.delegate -> right
+            else -> BigDecimal(value)
+        }
     }
```

Comment:
I feel like this `coalesceOrCreate` might be a candidate for a generic function

Reply:
That'd be nice but our `BigInteger` and `BigDecimal` definitions don't have a common interface which defines `val delegate` as a public member. I'd rather not add one either since the delegate is an internal implementation detail.